Using eagerInitSingletons in an ApplicationContextConfigurer breaks injecting an httpclient into tests

### Issue description

Micronaut CRaC adds a context configurer to eagerly initialize singletons:

```java
    @ContextConfigurer
    public static class Configurer implements ApplicationContextConfigurer {
        @Override
        public void configure(@NonNull ApplicationContextBuilder builder) {
            builder.eagerInitSingletons(true);
        }
    }
```

However when this is applied, if you add a controller to a plain 3.8.4 app:

```java
package example.micronaut;

import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;

import java.util.Collections;
import java.util.Map;

@Controller 
public class HelloController {
    @Get
    public Map<String, Object> index() {
        return Collections.singletonMap("message", "Hello World"); 
    }
}
```

and a test: 

```java
package example.micronaut;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import io.micronaut.http.HttpRequest;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Test;

import jakarta.inject.Inject;
import org.junit.jupiter.api.Disabled;

@MicronautTest // <1>
public class HelloControllerTest {

    @Inject
    @Client("/")  // <2>
    HttpClient client;

    @Test
    public void testHello() {
        String body = client.toBlocking().retrieve(HttpRequest.GET("/"));
        assertNotNull(body);
        assertEquals("{\"message\":\"Hello World\"}", body);
    }
}

```

Test fails with:

```
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [example.micronaut.HelloControllerTest] could not be loaded: Error instantiating bean of type  [example.micronaut.HelloControllerTest]

Message: Retrieving the port from the server before it has started is not supported when binding to a random port
Path Taken: new HelloControllerTest() --> HelloControllerTest.client
	at app//io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1921)
	at app//io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:249)
	at app//io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3326)
	at app//io.micronaut.context.DefaultBeanContext.finalizeConfiguration(DefaultBeanContext.java:3684)
	at app//io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:341)
	at app//io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:194)
	at app//io.micronaut.test.extensions.AbstractMicronautExtension.startApplicationContext(AbstractMicronautExtension.java:433)
	at app//io.micronaut.test.extensions.AbstractMicronautExtension.beforeClass(AbstractMicronautExtension.java:314)
	at app//io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeAll(MicronautJunit5Extension.java:84)
	at app//org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$12(ClassBasedTestDescriptor.java:395)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:395)
	at app//org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:211)
	at app//org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:84)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base@19.0.2/java.util.ArrayList.forEach(ArrayList.java:1511)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at app//org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at app//org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base@19.0.2/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base@19.0.2/java.lang.reflect.Method.invoke(Method.java:578)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type  [example.micronaut.HelloControllerTest]
```

Micronaut Framework: 3.8.4

[bugcractest.zip](https://github.com/micronaut-projects/micronaut-crac/files/10722071/bugcractest.zip)

--- a/.github/workflows/graalvm.yml
+++ b/.github/workflows/graalvm.yml
@@ -42,6 +42,7 @@ jobs:
           version: ${{ matrix.graalvm }}
           java-version: ${{ matrix.java }}
           components: 'native-image'
+          github-token: ${{ secrets.GITHUB_TOKEN }}
       - name: Build with Gradle
         run: |
           if ./gradlew tasks --no-daemon --all | grep -w "testNativeImage"

--- a/src/main/docs/guide/introduction.adoc
+++ b/src/main/docs/guide/introduction.adoc
@@ -32,13 +32,39 @@ This is achieved through a set of annotations:
 
 These annotations use internal Micronaut features and do not mock any part of Micronaut itself. When you run a test within `@MicronautTest` it is running your real application.
 
-In some tests you may need a reference to the `ApplicationContext` and/or the `EmbeddedServer` (for example, to create an instance of an `HttpClient`). Rather than defining these as properties of the test (such as a `@Shared` property in Spock), when using `@MicronautTest` you can reference the server/context that was started up for you, and inject them directly in your test.  
+In some tests you may need a reference to the `ApplicationContext` and/or the `EmbeddedServer` (for example, to create an instance of an `HttpClient`). Rather than defining these as properties of the test (such as a `@Shared` property in Spock), when using `@MicronautTest` you can reference the server/context that was started up for you, and inject them directly in your test.
 
 [source,groovy]
 ----
-@Inject 
+@Inject
 EmbeddedServer server //refers to the server that was started up for this test suite
 
-@Inject 
+@Inject
 ApplicationContext context //refers to the current application context within the scope of the test
+----
 
+### Eager Singleton Initialization
+
+If you enable https://docs.micronaut.io/latest/guide/index.html#eagerInit[eager singleton initialization] in your application, Micronaut will eagerly initialize all singletons at startup time. This can be useful for applications that need to perform some initialization at startup time, such as registering a bean with a third party library.
+
+However, as tests annotated with `@MicronautTest` are implicitly in the `Singleton` scope, this can cause problems injecting some beans (for example an `HttpClient`) into your test class.
+
+To avoid this, you can either disable eager singleton initialization for your tests, or you will need to manually get an instance of the bean you would normally inject.  As an example, to get an `HttpClient` you could do:
+
+[source,java]
+----
+@Inject
+EmbeddedServer server; // <1>
+
+Supplier<HttpClient> client = SupplierUtil.memoizedNonEmpty(() ->
+    server.getApplicationContext().createBean(HttpClient.class, server.getURL())); // <2>
+
+@Test
+void testEagerSingleton() {
+    Assertions.assertEquals("eager", client.get().toBlocking().retrieve("/eager")); // <3>
+}
+----
+
+<1> Inject the `EmbeddedServer` as normal
+<2> Create a `Supplier` that will create the `HttpClient` when it is first called
+<3> Use the `Supplier` to get the `HttpClient` and make the request

--- a/src/main/docs/guide/introduction.adoc
+++ b/src/main/docs/guide/introduction.adoc
@@ -52,17 +52,9 @@ However, as tests annotated with `@MicronautTest` are implicitly in the `Singlet
 To avoid this, you can either disable eager singleton initialization for your tests, or you will need to manually get an instance of the bean you would normally inject.  As an example, to get an `HttpClient` you could do:
 
 [source,java]
+.Using an HttpClient in a test with eager singleton initialization enabled
 ----
-@Inject
-EmbeddedServer server; // <1>
-
-Supplier<HttpClient> client = SupplierUtil.memoizedNonEmpty(() ->
-    server.getApplicationContext().createBean(HttpClient.class, server.getURL())); // <2>
-
-@Test
-void testEagerSingleton() {
-    Assertions.assertEquals("eager", client.get().toBlocking().retrieve("/eager")); // <3>
-}
+include::test-junit5/src/test/java/io/micronaut/test/junit5/EagerInitializationTest.java[tags=eager, indent=0]
 ----
 
 <1> Inject the `EmbeddedServer` as normal

--- a/src/main/docs/guide/introduction.adoc
+++ b/src/main/docs/guide/introduction.adoc
@@ -45,7 +45,7 @@ ApplicationContext context //refers to the current application context within th
 
 ### Eager Singleton Initialization
 
-If you enable https://docs.micronaut.io/latest/guide/index.html#eagerInit[eager singleton initialization] in your application, Micronaut will eagerly initialize all singletons at startup time. This can be useful for applications that need to perform some initialization at startup time, such as registering a bean with a third party library.
+If you enable https://docs.micronaut.io/latest/guide/index.html#eagerInit[eager singleton initialization] in your application, the Micronaut Framework eagerly initializes all singletons at startup time. This can be useful for applications that need to perform some initialization at startup time, such as registering a bean with a third party library.
 
 However, as tests annotated with `@MicronautTest` are implicitly in the `Singleton` scope, this can cause problems injecting some beans (for example an `HttpClient`) into your test class.
 

--- a/test-junit5/src/test/java/io/micronaut/test/junit5/EagerInitializationTest.java
+++ b/test-junit5/src/test/java/io/micronaut/test/junit5/EagerInitializationTest.java
@@ -1,0 +1,55 @@
+package io.micronaut.test.junit5;
+
+import io.micronaut.context.ApplicationContextBuilder;
+import io.micronaut.context.ApplicationContextConfigurer;
+import io.micronaut.context.annotation.ContextConfigurer;
+import io.micronaut.context.annotation.Property;
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.util.SupplierUtil;
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Get;
+import io.micronaut.http.client.HttpClient;
+import io.micronaut.runtime.server.EmbeddedServer;
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
+import jakarta.inject.Inject;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.function.Supplier;
+
+@MicronautTest
+@Property(name = "spec.name", value = "EagerSingletonTest")
+class EagerSingletonTest {
+
+    // tag::eager[]
+    @Inject
+    EmbeddedServer server; // <1>
+
+    Supplier<HttpClient> client = SupplierUtil.memoizedNonEmpty(() ->
+        server.getApplicationContext().createBean(HttpClient.class, server.getURL())); // <2>
+
+    @Test
+    void testEagerSingleton() {
+        Assertions.assertEquals("eager", client.get().toBlocking().retrieve("/eager")); // <3>
+    }
+    // end::eager[]
+
+    @Requires(property = "spec.name", value = "EagerSingletonTest")
+    @ContextConfigurer
+    public static class Configurer implements ApplicationContextConfigurer {
+        @Override
+        public void configure(@NonNull ApplicationContextBuilder builder) {
+            builder.eagerInitSingletons(true);
+        }
+    }
+
+    @Requires(property = "spec.name", value = "EagerSingletonTest")
+    @Controller("/eager")
+    public static class EagerController {
+        @Get
+        String test() {
+            return "eager";
+        }
+    }
+}

--- a/test-junit5/src/test/java/io/micronaut/test/junit5/EagerInitializationTest.java
+++ b/test-junit5/src/test/java/io/micronaut/test/junit5/EagerInitializationTest.java
@@ -18,7 +18,6 @@ import java.util.function.Supplier;
 
 @MicronautTest(contextBuilder = EagerInitializationTest.EagerContextBuilder.class)
 @Property(name = "spec.name", value = "EagerInitializationTest")
-@Property(name = "micronaut.eager-init-singletons", value = "true")
 class EagerInitializationTest {
 
     // tag::eager[]

--- a/test-junit5/src/test/java/io/micronaut/test/junit5/EagerInitializationTest.java
+++ b/test-junit5/src/test/java/io/micronaut/test/junit5/EagerInitializationTest.java
@@ -1,11 +1,9 @@
 package io.micronaut.test.junit5;
 
-import io.micronaut.context.ApplicationContextBuilder;
-import io.micronaut.context.ApplicationContextConfigurer;
-import io.micronaut.context.annotation.ContextConfigurer;
+import io.micronaut.context.DefaultApplicationContextBuilder;
 import io.micronaut.context.annotation.Property;
 import io.micronaut.context.annotation.Requires;
-import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Introspected;
 import io.micronaut.core.util.SupplierUtil;
 import io.micronaut.http.annotation.Controller;
 import io.micronaut.http.annotation.Get;
@@ -18,9 +16,10 @@ import org.junit.jupiter.api.Test;
 
 import java.util.function.Supplier;
 
-@MicronautTest
-@Property(name = "spec.name", value = "EagerSingletonTest")
-class EagerSingletonTest {
+@MicronautTest(contextBuilder = EagerInitializationTest.EagerContextBuilder.class)
+@Property(name = "spec.name", value = "EagerInitializationTest")
+@Property(name = "micronaut.eager-init-singletons", value = "true")
+class EagerInitializationTest {
 
     // tag::eager[]
     @Inject
@@ -35,21 +34,19 @@ class EagerSingletonTest {
     }
     // end::eager[]
 
-    @Requires(property = "spec.name", value = "EagerSingletonTest")
-    @ContextConfigurer
-    public static class Configurer implements ApplicationContextConfigurer {
-        @Override
-        public void configure(@NonNull ApplicationContextBuilder builder) {
-            builder.eagerInitSingletons(true);
-        }
-    }
-
-    @Requires(property = "spec.name", value = "EagerSingletonTest")
+    @Requires(property = "spec.name", value = "EagerInitializationTest")
     @Controller("/eager")
     public static class EagerController {
         @Get
         String test() {
             return "eager";
+        }
+    }
+
+    @Introspected
+    static class EagerContextBuilder extends DefaultApplicationContextBuilder {
+        public EagerContextBuilder() {
+            eagerInitSingletons(true);
         }
     }
 }

